ecosystem/typescript/sdk/src/plugins/ans_client.ts

```diff
@@ -0,0 +1,175 @@
+import { Provider } from "../providers";
+import * as Gen from "../generated/index";
+
+const ansContractsMap: Record<string, string> = {
+  testnet: "0x5f8fd2347449685cf41d4db97926ec3a096eaf381332be4f1318ad4d16a8497c",
+  mainnet: "0x867ed1f6bf916171b1de3ee92849b8978b7d1b9e0a8cc982a3d19d535dfd9c0c",
+};
+
+// Each name component can only have lowercase letters, number or hyphens, and cannot start or end with a hyphen.
+const nameComponentPattern = /^[a-z\d][a-z\d-]{1,61}[a-z\d]$/;
+
+const namePattern = new RegExp(
+  "^" +
+    // Optional subdomain (cannot be followed by .apt)
+    "(?:(?<subdomain>[^.]+)\\.(?!apt$))?" +
+    // Domain
+    "(?<domain>[^.]+)" +
+    // Optional .apt suffix
+    "(?:\\.apt)?" +
+    "$",
+);
+
+type ReverseLookupRegistryV1 = {
+  registry: {
+    handle: string;
+  };
+};
+
+type NameRegistryV1 = {
+  registry: {
+    handle: string;
+  };
+};
+
+export class AnsClient {
+  contractAddress: string;
+
+  provider: Provider;
+
+  /**
+   * Creates new AnsClient instance
+   * @param provider Provider instance
+   * @param contractAddress An optional contract address.
+   * If there is no contract address matching to the provided network
+   * then the AnsClient class expects a contract address -
+   * this is to support both mainnet/testnet networks and local development.
+   */
+  constructor(provider: Provider, contractAddress?: string) {
+    this.provider = provider;
+    if (!ansContractsMap[this.provider.network] && !contractAddress) {
+      throw new Error("Please provide a valid contract address");
```

Comment:
Perhaps include what the valid options are in this error message.

Reply:
this is for when they dont provide a contract address, maybe - `Please provide a contract address` is better.

Reply:
Well this is for both right, either network or address? Maybe separate the two into two separate checks + errors?

Reply:
hmm so this is for when `Provider` was passed with no `testnet / mainnet` network && no `contractAddress`. There is now error for not passing in a custom provider, like
```
new Provider({ fullnodeUrl: "full-node-url", indexerUrl: "indexer-url" });
```
but if you do, you must provide a contract address. check out https://github.com/aptos-labs/aptos-core/pull/7227/commits/f78b767b3df2d58751b6c6f6a21c98fd97407b24#diff-8f89caf6aef6a55235027e6ec2973415e27b57af2ffdceedfc73d5087dbeb4ba

Reply:
Ohhhhhhh I'm dumb, I was misreading the code. Okay yeah this makes sense. Maybe a different message like "Error: For custom providers, you must pass in a contract address".